context.go

```diff
@@ -0,0 +1,143 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build (linux || darwin) && (amd64 || arm64)
+
+package waf
+
+import (
+	"sync"
+	"time"
+
+	"go.uber.org/atomic"
+)
+
+// Context is a WAF execution context.
+// It's a wrapper over the void * the WAF returns with ddwaf_context_init()
+// We store in there:
+// * A mutex to order calls to ddwaf_run()
+// * The handle it was created with
+// * cgoRefs to keep references to the go waf objects we sent
+// * A number of metrics
+type Context struct {
+	// Instance of the WAF
+	handle   *Handle
+	cContext wafContext
+
+	// Mutex protecting the use of context which is not thread-safe.
+	mutex sync.Mutex
+
+	// cgoRefs is used to retain go references to WafObjects until the context is destroyed.
+	// As per libddwaf documentation, WAF Objects must be alive during all the context lifetime
+	cgoRefs cgoRefPool
+
+	// Stats
+	// Cumulated internal WAF run time - in nanoseconds - for this context.
+	totalRuntimeNs atomic.Uint64
+	// Cumulated overall run time - in nanoseconds - for this context.
+	totalOverallRuntimeNs atomic.Uint64
+	// Cumulated timeout count for this context.
+	timeoutCount atomic.Uint64
+}
+
+// Run encodes the input data to send it to the WAF and returns the return values of ddwaf_run()
+// The parameters are the following:
+// * addressesToData: a map of key WAF addresses to the user input. The user input can be any go type.
+// * timeout: a timeout value to run the waf. If the threshold is exceeded, ddwaf_run will return without completing everything
+// The return values are the following:
+// * matches: A json string directly sent by the WAF containing all the data necessary for the backend
+// * actions: A array of string referencing actions from triggered rules (`on_match` rule field)
+// * err: a RunError error from the error codes the WAF returns
```

Comment:
Making it a bit higher level and better looking for godoc:

```suggestion
// Run encodes the given addressesToData values and runs them against the WAF rules within the given
// timeout value. It returns the matches as a JSON string (usually opaquely used) along with the corresponding
// actions in any. In case of an error, matches and actions can still be returned, for instance in the case of a
// timeout error. Errors can be tested against the RunError type.
```